python/fusion_engine_client/analysis/analyzer.py

```diff
@@ -340,6 +340,126 @@ def _calc_stats(input):
 
         self._add_figure(name="time_scale", figure=figure, title="Time Scale")
 
+    def plot_reset_timing(self):
+        if self.output_dir is None:
+            return
+
+        # Find reset events.
+        result = self.reader.read(message_types=[EventNotificationMessage], return_message_index=True, **self.params)
+        event_data = result[EventNotificationMessage.MESSAGE_TYPE]
+
+        reset_idx = event_data.event_type == EventType.RESET
+        if not np.any(reset_idx):
+            self.logger.info('No reset events detected. Skipping reset timing type plot.')
+            return
+
+        self.logger.info('Calculating reset recovery times...')
+
+        # Note that events contain system time, not P1 time. We'll assume system time is close enough to P1 time for
+        # purposes of calculating elapsed reset time below. In the future, we'll have a mechanism for accurately
+        # converting between system and P1 time.
+        reset_system_time_sec = event_data.system_time[reset_idx]
+
+        reset_idx = np.where(reset_idx)[0]
+        reset_message_indices = [event_data.message_index[i] for i in reset_idx]
+
+        # For each reset in the log, try to find the pose messages immediately following the reset where the solution
+        # type first goes invalid, and then where it goes valid again.
+        dt_reset_to_valid = np.full(reset_idx.shape, np.nan)
+        dt_reset_to_invalid = np.full(reset_idx.shape, np.nan)
+        dt_invalid_to_valid = np.full(reset_idx.shape, np.nan)
+        unstarted_resets = []
+
+        log_reader = self.reader.get_log_reader()
+        for i, reset_index in enumerate(reset_message_indices):
+            next_reset_index = reset_message_indices[i + 1] if i < len(reset_message_indices) - 1 else None
+
+            # Filter to all pose messages _after_ the reset event.
+            log_reader.clear_filters()
+            log_reader.rewind()
+            log_reader.filter_in_place(slice(reset_index + 1, next_reset_index, 1))
+            log_reader.filter_in_place(self.params['time_range'])
+            log_reader.filter_in_place(PoseMessage)
+            log_reader.set_show_progress(False)
+
+            # Find the pose where the solution went invalid after the reset, then where it went valid after that.
+            invalid_p1_time = None
+            valid_p1_time = None
+            while True:
+                try:
+                    _, message, pose_index = self.reader.read_next(return_message_index=True)
+                except StopIteration:
+                    break
+
+                if next_reset_index is not None and pose_index >= next_reset_index:
```

Comment:
Is this condition possible since you've already filtered the log to not include times past `next_reset_index`?

Reply:
Nope don't think so. I added this before I added the slice filtering above. Will remove.